File: CS/CustomDatesPivot/DateDataSourceWrapper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Collections;

namespace CustomDatesPivot
{
    class DateDataSourceWrapper : ITypedList, IList, IEnumerator
    {
        int position = -1;
        public readonly IList NestedList;
        public readonly List<DateTime> CustomDates;
        public ITypedList NestedTypedList { get { return (ITypedList)NestedList; } }
        public DateDataSourceWrapper(ITypedList OriginalDatasource, List<string> dateFields, List<DateTime> customDates)
        {
            _DateFields = dateFields;
            this.NestedList = (IList)OriginalDatasource;
            CustomDates = customDates;
        }

        private readonly List<string> _DateFields;



        class EmptyObjectPropertyDescriptor : PropertyDescriptor
        {
            private readonly List<string> _DateFields;
            public readonly PropertyDescriptor NestedDescriptor;
            public EmptyObjectPropertyDescriptor(PropertyDescriptor nestedDescriptor, List<string> dateFields)
                : base(nestedDescriptor.Name, (Attribute[])new ArrayList(nestedDescriptor.Attributes).ToArray(typeof(Attribute)))
            {
                _DateFields = dateFields;
                this.NestedDescriptor = nestedDescriptor;
            }
            public override bool CanResetValue(object component)
            {
                return false;
            }
            public override Type ComponentType
            {
                get { return typeof(object); }
            }
            public override object GetValue(object component)
            {
                if (component is DateTime)
                {
                    if (_DateFields.Contains(NestedDescriptor.Name))
                        return (DateTime)component;
                    return null;
                }
                else
                    return NestedDescriptor.GetValue(component);
            }
            public override bool IsReadOnly
            {
                get { return true; }
            }
            public override Type PropertyType
            {
                get { return NestedDescriptor.PropertyType; }
            }
            public override void ResetValue(object component)
            {
                throw new NotSupportedException("The method or operation is not implemented.");
            }
            public override void SetValue(object component, object value)
            {
                throw new NotSupportedException("The method or operation is not implemented.");
            }
            public override bool ShouldSerializeValue(object component)
            {
                return true;
            }
        }
        public PropertyDescriptorCollection GetItemProperties(PropertyDescriptor[] listAccessors)
        {
            List<PropertyDescriptor> result = new List<PropertyDescriptor>();
            foreach (PropertyDescriptor pd in NestedTypedList.GetItemProperties(ExtractOriginalDescriptors(listAccessors)))
            {

                result.Add(new EmptyObjectPropertyDescriptor(pd, _DateFields));
            }
            return new PropertyDescriptorCollection(result.ToArray());
        }
        public string GetListName(PropertyDescriptor[] listAccessors)
        {
            return NestedTypedList.GetListName(ExtractOriginalDescriptors(listAccessors));
        }

        protected static PropertyDescriptor[] ExtractOriginalDescriptors(PropertyDescriptor[] listAccessors)
        {
            if (listAccessors == null)
                return null;
            PropertyDescriptor[] convertedDescriptors = new PropertyDescriptor[listAccessors.Length];
            for (int i = 0; i < convertedDescriptors.Length; ++i)
            {
                PropertyDescriptor d = listAccessors[i];
                EmptyObjectPropertyDescriptor c = d as EmptyObjectPropertyDescriptor;
                if (c != null)
                    convertedDescriptors[i] = c.NestedDescriptor;
                else
                    convertedDescriptors[i] = d;
            }
            return convertedDescriptors;
        }
        public int Add(object value)
        {
            throw new NotSupportedException("The method or operation is not implemented.");
        }
        public void Clear()
        {
            throw new NotSupportedException("The method or operation is not implemented.");
        }
        public bool Contains(object value)
        {
            if (value is DateTime)
                return CustomDates.Contains((DateTime)value);
            return NestedList.Contains(value);
        }
        public int IndexOf(object value)
        {
            if (value is DateTime && CustomDates.Contains((DateTime)value))
                return CustomDates.IndexOf((DateTime)value);
            int nres = NestedList.IndexOf(value);
            if (nres < 0)
                return nres;
            return nres + CustomDates.Count;
        }
        public void Insert(int index, object value)
        {
            throw new NotSupportedException("The method or operation is not implemented.");
        }
        public bool IsFixedSize
        {
            get { return true; }
        }
        public bool IsReadOnly
        {
            get { return true; }
        }
        public void Remove(object value)
        {
            throw new NotSupportedException("The method or operation is not implemented.");
        }
        public void RemoveAt(int index)
        {
            throw new NotSupportedException("The method or operation is not implemented.");
        }
        public object this[int index]
        {
            get
            {
                if (index < CustomDates.Count)
                    return CustomDates[index];
                else
                    return NestedList[index - CustomDates.Count];
            }
            set
            {
                throw new NotSupportedException("The method or operation is not implemented.");
            }
        }
        public void CopyTo(Array array, int index)
        {
            throw new NotSupportedException("The method or operation is not implemented.");
        }
        public int Count
        {
            get { return NestedList.Count + CustomDates.Count; }
        }
        public bool IsSynchronized
        {
            get { return false; }
        }
        public object SyncRoot
        {
            get { return NestedList.SyncRoot; }
        }

        public object Current
        {
            get
            {
                if (position < CustomDates.Count)
                    return CustomDates[position];
                else
                    return NestedList[position - CustomDates.Count];
            }
        }

        public IEnumerator GetEnumerator()
        {
            return (IEnumerator)this;
        }

        public bool MoveNext()
        {
            position++;
            return (position < Count);
        }

        public void Reset()
        {
            position = 0;
        }
    }
}

File: VB/CustomDatesPivot/DateDataSourceWrapper.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Collections

Namespace CustomDatesPivot

    Friend Class DateDataSourceWrapper
        Implements ITypedList, IList

        Public ReadOnly NestedList As IList

        Public ReadOnly CustomDates As List(Of Date)

        Public ReadOnly Property NestedTypedList As ITypedList
            Get
                Return CType(NestedList, ITypedList)
            End Get
        End Property

        Public Sub New(ByVal OriginalDatasource As ITypedList, ByVal dateFields As List(Of String), ByVal customDates As List(Of Date))
            _DateFields = dateFields
            NestedList = CType(OriginalDatasource, IList)
            Me.CustomDates = customDates
        End Sub

        Private ReadOnly _DateFields As List(Of String)

        Private Class EmptyObjectPropertyDescriptor
            Inherits PropertyDescriptor

            Private ReadOnly _DateFields As List(Of String)

            Public ReadOnly NestedDescriptor As PropertyDescriptor

            Public Sub New(ByVal nestedDescriptor As PropertyDescriptor, ByVal dateFields As List(Of String))
                MyBase.New(nestedDescriptor.Name, CType(New ArrayList(nestedDescriptor.Attributes).ToArray(GetType(Attribute)), Attribute()))
                _DateFields = dateFields
                Me.NestedDescriptor = nestedDescriptor
            End Sub

            Public Overrides Function CanResetValue(ByVal component As Object) As Boolean
                Return False
            End Function

            Public Overrides ReadOnly Property ComponentType As Type
                Get
                    Return GetType(Object)
                End Get
            End Property

            Public Overrides Function GetValue(ByVal component As Object) As Object
                If TypeOf component Is Date Then
                    If _DateFields.Contains(NestedDescriptor.Name) Then Return CDate(component)
                    Return Nothing
                Else
                    Return NestedDescriptor.GetValue(component)
                End If
            End Function

            Public Overrides ReadOnly Property IsReadOnly As Boolean
                Get
                    Return True
                End Get
            End Property

            Public Overrides ReadOnly Property PropertyType As Type
                Get
                    Return NestedDescriptor.PropertyType
                End Get
            End Property

            Public Overrides Sub ResetValue(ByVal component As Object)
                Throw New NotSupportedException("The method or operation is not implemented.")
            End Sub

            Public Overrides Sub SetValue(ByVal component As Object, ByVal value As Object)
                Throw New NotSupportedException("The method or operation is not implemented.")
            End Sub

            Public Overrides Function ShouldSerializeValue(ByVal component As Object) As Boolean
                Return True
            End Function
        End Class

        Public Function GetItemProperties(ByVal listAccessors As PropertyDescriptor()) As PropertyDescriptorCollection Implements ITypedList.GetItemProperties
            Dim result As List(Of PropertyDescriptor) = New List(Of PropertyDescriptor)()
            For Each pd As PropertyDescriptor In NestedTypedList.GetItemProperties(ExtractOriginalDescriptors(listAccessors))
                result.Add(New EmptyObjectPropertyDescriptor(pd, _DateFields))
            Next

            Return New PropertyDescriptorCollection(result.ToArray())
        End Function

        Public Function GetListName(ByVal listAccessors As PropertyDescriptor()) As String Implements ITypedList.GetListName
            Return NestedTypedList.GetListName(ExtractOriginalDescriptors(listAccessors))
        End Function

        Protected Shared Function ExtractOriginalDescriptors(ByVal listAccessors As PropertyDescriptor()) As PropertyDescriptor()
            If listAccessors Is Nothing Then Return Nothing
            Dim convertedDescriptors As PropertyDescriptor() = New PropertyDescriptor(listAccessors.Length - 1) {}
            Dim i As Integer = 0
            While i < convertedDescriptors.Length
                Dim d As PropertyDescriptor = listAccessors(i)
                Dim c As EmptyObjectPropertyDescriptor = TryCast(d, EmptyObjectPropertyDescriptor)
                If c IsNot Nothing Then
                    convertedDescriptors(i) = c.NestedDescriptor
                Else
                    convertedDescriptors(i) = d
                End If

                Threading.Interlocked.Increment(i)
            End While

            Return convertedDescriptors
        End Function

        Public Function Add(ByVal value As Object) As Integer Implements IList.Add
            Throw New NotSupportedException("The method or operation is not implemented.")
        End Function

        Public Sub Clear() Implements IList.Clear
            Throw New NotSupportedException("The method or operation is not implemented.")
        End Sub

        Public Function Contains(ByVal value As Object) As Boolean Implements IList.Contains
            If TypeOf value Is Date Then Return CustomDates.Contains(CDate(value))
            Return NestedList.Contains(value)
        End Function

        Public Function IndexOf(ByVal value As Object) As Integer Implements IList.IndexOf
            If TypeOf value Is Date AndAlso CustomDates.Contains(CDate(value)) Then Return CustomDates.IndexOf(CDate(value))
            Dim nres As Integer = NestedList.IndexOf(value)
            If nres < 0 Then Return nres
            Return nres + CustomDates.Count
        End Function

        Public Sub Insert(ByVal index As Integer, ByVal value As Object) Implements IList.Insert
            Throw New NotSupportedException("The method or operation is not implemented.")
        End Sub

        Public ReadOnly Property IsFixedSize As Boolean Implements IList.IsFixedSize
            Get
                Return True
            End Get
        End Property

        Public ReadOnly Property IsReadOnly As Boolean Implements IList.IsReadOnly
            Get
                Return True
            End Get
        End Property

        Public Sub Remove(ByVal value As Object) Implements IList.Remove
            Throw New NotSupportedException("The method or operation is not implemented.")
        End Sub

        Public Sub RemoveAt(ByVal index As Integer) Implements IList.RemoveAt
            Throw New NotSupportedException("The method or operation is not implemented.")
        End Sub

        Default Public Property Item(ByVal index As Integer) As Object Implements IList.Item
            Get
                If index < CustomDates.Count Then
                    Return CustomDates(index)
                Else
                    Return NestedList(index - CustomDates.Count)
                End If
            End Get

            Set(ByVal value As Object)
                Throw New NotSupportedException("The method or operation is not implemented.")
            End Set
        End Property

        Public Sub CopyTo(ByVal array As Array, ByVal index As Integer) Implements ICollection.CopyTo
            Throw New NotSupportedException("The method or operation is not implemented.")
        End Sub

        Public ReadOnly Property Count As Integer Implements ICollection.Count
            Get
                Return NestedList.Count + CustomDates.Count
            End Get
        End Property

        Public ReadOnly Property IsSynchronized As Boolean Implements ICollection.IsSynchronized
            Get
                Return False
            End Get
        End Property

        Public ReadOnly Property SyncRoot As Object Implements ICollection.SyncRoot
            Get
                Return NestedList.SyncRoot
            End Get
        End Property

        Public Function GetEnumerator() As IEnumerator Implements IEnumerable.GetEnumerator
            Throw New NotSupportedException("The method or operation is not implemented.")
        End Function
    End Class
End Namespace
